sdk/include/opentelemetry/sdk/trace/sampler.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include "opentelemetry/common/attribute_value.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/common/atomic_shared_ptr.h"
+#include "opentelemetry/trace/span.h"
+#include "opentelemetry/trace/trace_id.h"
+#include "opentelemetry/trace/tracer.h"
+#include "opentelemetry/version.h"
+
+#include <map>
+#include <memory>
+#include <string>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+class Sampler
+{
+private:
+  // Placeholder
+  class SpanContext
+  {};
+
+public:
+  virtual ~Sampler() = default;
+  /**
+   * Called during Span creation to make a sampling decision.
+   *
+   * @param parent_context TODO: a shared pointer of the SpanContext of a parent Span. 
+   *     null if this is a root span.
+   * @param trace_id the TraceId for the new Span. This will be identical to that in
+   *     the parentContext, unless this is a root span.
+   * @param name the name of the new Span.
+   * @param spanKind the trace_api::SpanKind of the Span.
+   * @param attributes TODO: current map is a placeholder.
+   *     list of AttributeValue with their keys.
+   * @param links TODO: Collection of links that will be associated with the Span to be created.
+   * @return sampling result whether span should be sampled or not.
+   * @since 0.1.0
+   */
+
+  virtual SamplingResult ShouldSample(
```

Comment:
You cannot return an abstract class. Consider making this a `std::unique_ptr<SamplingResult>`.

Reply:
Can we have a non-abstract SamplingResult so we don't force a heap allocation here?

Reply:
I also think it make sense to use non-abstract SamplingResult bacause the SamplingResult for all built-in samplers requires the same functionality.

Reply:
Agreed. What do you think about making this a single struct:
```c++
struct SamplingResult {
  Decision decision;
  std::unique_ptr<std::map<std::string, common::AttributeValue>> attributes;
};
```
This would also be pretty efficient for simple samplers that don't return any attributes (no heap allocations involved).

We could use STL types here, as there are no ABI stability requirements for the SDK.

Reply:
It looks good to me. Since `SamplingResult` struct only takes 16 bytes in 64-bit machine, `ShouldSample` can return `SamplingResult` directly right?